### PR TITLE
Wait for runtime to be ready in __init__

### PR DESCRIPTION
**Pull request by rbren**
**Short description of the problem this fixes or functionality that this introduces. This may be used for the CHANGELOG**

I fixed this a few days ago, but my fix seems to have disappeared.

---
**Give a summary of what the PR does, explaining any non-trivial design decisions**

* Wait for client in __init__
* Refactor the `wait_for_client` function so that log refreshing is done separately, making it easier for other callers to use

---
**Link of any specific issues this addresses**

### CODE REVIEW

**Comment by tofarr** [comment]
Looks like it was removed in this commit: https://github.com/All-Hands-AI/OpenHands/commit/a45b20a406bed9f5951cf26dc2844373e8230737#diff-a7ff0449d091e3e762a4a8b604fe16a3a5570448751817a5a3e207ab18ec648c

I'd like to know if the removal was accidental before putting this back in. @tobitege - any insights?

**Comment by tobitege** [comment]
> Looks like the call to `_wait_until_alive` was removed in this commit: [a45b20a#diff-a7ff0449d091e3e762a4a8b604fe16a3a5570448751817a5a3e207ab18ec648c](https://github.com/All-Hands-AI/OpenHands/commit/a45b20a406bed9f5951cf26dc2844373e8230737#diff-a7ff0449d091e3e762a4a8b604fe16a3a5570448751817a5a3e207ab18ec648c)
> 
> I'd like to know if the removal was accidental before putting this back in. @tobitege - any insights?

Those 2 commits are not removing _your_ call, but comment out a log refresh in `run_action` from a previous PR.

**Comment by tobitege** [comment]
Line 162 adding back makes sense. If I removed it, then by accident.
But the other changes that process the logs should not just be changed like that, they're there to wait for the client become ready on first call.

**Comment by rbren** [comment]
But since it's alive in `__init__`, can't we assume it's alive everywhere else?

**Comment by tobitege** [comment]
> But since it's alive in `__init__`, can't we assume it's alive everywhere else?

Still, the very first call of _wait_until_alive because of potential timing issues should make sure it got'em all and that extra check is only done once.

**Comment by tobitege** [comment]
in #3962 I just tested with minimal reverts, that sudo works again

File: openhands/runtime/client/runtime.py [hunk 1]
```diff
@@ -423,6 +418,7 @@ def run_action(self, action: Action) -> Observation:
                 )
                 if response.status_code == 200:
                     output = response.json()
+                    logger.info(f'Response content: {output}')
```
Comment (rbren) [inline openhands/runtime/client/runtime.py:421]:
```suggestion
```

**Comment by rbren** [comment]
@tobitege sorry can you explain more? I assume you're correct given that the tests are failing, but once `/alive` returns 200, can't we assume it's alive from that point forward?

**Comment by tobitege** [comment]
> @tobitege sorry can you explain more? I assume you're correct given that the tests are failing, but once `/alive` returns 200, can't we assume it's alive from that point forward?

Oops, I had added this to my previous comment instead of making a new one:
The websocket becomes alive but we have to wait for the client to be ready (bash/ipython) before we can start sending commands.

**Comment by rbren** [comment]
Thanks for the explanation in Slack Tobi!

I think I have a better fix now--`_wait_until_alive` will now always wait for `log_buffer.client_ready`. We also reuse the same retry logic instead of having separate retry logic for the log checker

**Comment by tobitege** [comment]
> Thanks for the explanation in Slack Tobi!
> 
> I think I have a better fix now--`_wait_until_alive` will now always wait for `log_buffer.client_ready`. We also reuse the same retry logic instead of having separate retry logic for the log checker

Yes, looks much cleaner!

**Comment by rbren** [comment]
OK @tobitege I think the tests should finally be passing

**Comment by tobitege** [comment]
> OK @tobitege I think the tests should finally be passing

Ok. But why move the wait before the `super.__init__` in the latest commit? That is doing the plugin initialization.

**Comment by rbren** [comment]
The super __init__ sets env vars, so it needs the environment to be ready

**Comment by tobitege** [comment]
> The super **init** sets env vars, so it needs the environment to be ready. Otherwise the tests fail

Hmm ok, due to the `@retry` "polling" effect now. 

Any thoughts on the wait times, though?

**Comment by rbren** [comment]
OK I refactored a bit so that the env is set later on--hopefully this works

**Review by tobitege — APPROVED** [review]
LGTM

**Pull request merged**